pkg/actors/actors.go

```diff
@@ -15,1431 +15,554 @@ package actors
+	rch := a.registerDoneCh
+	a.registerDoneLock.RUnlock()
 
-func (a *actorsRuntime) GetReminder(ctx context.Context, req *GetReminderRequest) (*internal.Reminder, error) {
-	if !a.actorsConfig.Config.HostedActorTypes.IsActorTypeHosted(req.ActorType) {
-		return nil, ErrReminderOpActorNotHosted
+	select {
+	case <-rch:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
 	}
-
-	return a.actorsReminders.GetReminder(ctx, req)
-}
-
-func (a *actorsRuntime) DeleteTimer(ctx context.Context, req *DeleteTimerRequest) error {
-	return a.timers.DeleteTimer(ctx, req.Key())
 }
 
-func (a *actorsRuntime) RegisterInternalActor(ctx context.Context, actorType string, factory InternalActorFactory, actorIdleTimeout time.Duration) error {
-	if !a.haveCompatibleStorage() {
-		return fmt.Errorf("unable to register internal actor type '%s': %w", actorType, ErrIncompatibleStateStore)
+func (a *actors) handleIdleActor(target targets.Idlable) {
+	if err := a.placement.Lock(context.Background()); err != nil {
+		return
 	}
+	defer a.placement.Unlock()
 
-	// Call GetOrSet which returns "existing=true" if the actor type was already registered
-	_, existing := a.internalActorTypes.GetOrSet(actorType, factory)
-	if existing {
-		return fmt.Errorf("actor type '%s' already registered", actorType)
-	}
-
-	log.Debugf("Registered internal actor type '%s'", actorType)
+	log.Debugf("Actor %s is idle, deactivating", target.Key())
 
-	a.actorsConfig.Config.HostedActorTypes.AddActorType(actorType, actorIdleTimeout)
-
-	if a.placementEnabled {
-		err := a.placement.AddHostedActorType(actorType, actorIdleTimeout)
-		if err != nil {
-			return fmt.Errorf("error updating hosted actor types: %w", err)
-		}
+	if err := a.table.Halt(context.Background(), target); err != nil {
+		log.Errorf("Failed to halt actor %s: %s", target.Key(), err)
+		return
 	}
-	return nil
 }
 
-func (a *actorsRuntime) GetRuntimeStatus(ctx context.Context) *runtimev1pb.ActorRuntime {
-	// Do not populate RuntimeStatus, which will be populated by the runtime
-	res := &runtimev1pb.ActorRuntime{
-		ActiveActors: a.getActiveActorsCount(ctx),
+func (a *actors) RuntimeStatus() *runtimev1pb.ActorRuntime {
+	select {
+	case <-a.registerDoneCh:
+		if a.disabled.Load() != nil {
+			return &runtimev1pb.ActorRuntime{
+				RuntimeStatus: runtimev1pb.ActorRuntime_DISABLED,
+				Placement:     "placement: disconnected",
+			}
+		}
+	default:
+		if len(a.placementAddresses) == 0 {
+			return &runtimev1pb.ActorRuntime{
+				Placement:     "placement: disconnected",
+				RuntimeStatus: runtimev1pb.ActorRuntime_DISABLED,
+			}
+		} else {
+			return &runtimev1pb.ActorRuntime{
+				RuntimeStatus: runtimev1pb.ActorRuntime_INITIALIZING,
+			}
+		}
 	}
 
-	if a.placementEnabled {
-		res.HostReady = a.placement.PlacementHealthy() && a.haveCompatibleStorage()
-		res.Placement = a.placement.StatusMessage()
+	hostReady := true
+	statusMessage := "placement: connected"
+	if !a.placement.Ready() {
+		statusMessage = "placement: disconnected"
```

Comment:
This is a nit, but `"placement: disconnected"` is used 3 times in this file as a magic string. Maybe we should pull it out to a const?